### PR TITLE
Merge 3.3 to 3.4

### PR DESCRIPTION
**Pull request by Aflynn50**
There were minor conflicts with version numbers in various places and the upgrade steps #16969 had to be updated to target upgrades through 3.4.1. The upgrade steps present for 3.3.1 that had made there way in here by accident were removed on the advice of @wallyworld.

 The PRs included in this merge up are:
- #16999 from Aflynn50
- #16997 from wallyworld
- #16969 from jack-w-shaw
- #16988 from tlm
- #16983 from Aflynn50
- #16973 from Aflynn50
- #16980 from juju
- #16976 from Aflynn50
- #16974 from wallyworld
- #16972 from jack-w-shaw
- #16970 from tlm
- #16959 from juju
- #16957 from jameinel
- #16932 from hpidcock
- #16945 from benhoyt
- #16943 from wallyworld

### CODE REVIEW

File: upgrades/operations.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ import (
 var stateUpgradeOperations = func() []Operation {
 	steps := []Operation{
 		upgradeToVersion{version.MustParse("3.3.1"), stateStepsFor331()},
+		upgradeToVersion{version.MustParse("3.3.4"), stateStepsFor334()},
```
Comment (jack-w-shaw) [inline upgrades/operations.go:23]: Uh oh

Version here needs a change

Looks like someone else made the same mistake though previously. We should investigate ConvertApplicationsOfferTokenKeys and fix before merging this

**Review by jack-w-shaw — CHANGES_REQUESTED** [review]

File: upgrades/operations.go (hunk 1 above)
Comment (Aflynn50) [reply upgrades/operations.go]: Changed the version and updated the tests.

File: upgrades/backend.go [hunk 2]
```diff
@@ -12,6 +12,7 @@ import (
 // StateBackend provides an interface for upgrading the global state database.
 type StateBackend interface {
 	ConvertApplicationOfferTokenKeys() error
```
Comment (jack-w-shaw) [inline upgrades/backend.go:14]: Needs to be dropped here as well

File: state/upgrades.go [hunk 3]
```diff
@@ -143,3 +144,39 @@ func ConvertApplicationOfferTokenKeys(pool *StatePool) error {
 		return nil
```
Comment (jack-w-shaw) [inline state/upgrades.go:144]: ^ Up here as welll

(I can't comment too far away from the change

File: state/upgrades_test.go [hunk 4]
```diff
@@ -154,3 +154,104 @@ func (s *upgradesSuite) TestConvertApplicationOfferTokenKeys(c *gc.C) {
 	expectedData := upgradedData(remoteEntitiesColl, expected)
```
Comment (jack-w-shaw) [inline state/upgrades_test.go:154]: ^ and here

File: upgrades/backend.go [hunk 5]
```diff
@@ -33,3 +34,7 @@ type stateBackend struct {
 func (s stateBackend) ConvertApplicationOfferTokenKeys() error {
 	return state.ConvertApplicationOfferTokenKeys(s.pool)
 }
```
Comment (jack-w-shaw) [inline upgrades/backend.go:36]: One more (hopefully)

Comment (Aflynn50) [reply upgrades/backend.go]: Its not my day

**Review by jack-w-shaw — APPROVED** [review]
My parts look good

**Pull request merged**